File: Project Euler/0029 - Distinct powers.cpp
#include <bits/stdc++.h> 
using namespace std; 
    
int main()
{
    int n = 1000000;
    bitset<1000001> A;
    int maxPow = (int) (log(n)/log(2)) + 1;
    long pows [maxPow]={};
    for (int i=2; i<=n; i++){
        long a = i;
        for (int j=1; j<=n; j++){
            if(!A[a]){
                A[a] = true;
                pows[j] += 1;
            } else {
                break;
            }
            a = a*i;
            if (a>n) {
                break;
            }
        }
    }
    bitset<20000001> allPows;
    long total = 0;
    int previous = 0;
    for (int i=1; i < maxPow; i++){
        for  (int j=2; j<=n; j++){
            int prod = i*j;
            allPows[prod] = true;
        }
        total = total + ( pows[i] * (allPows.count() - previous));
        previous = allPows.count();
    }
    cout<<total;
    return 0; 
}
//std::ios_base::sync_with_stdio(false);

File: Project Euler/0029 - Distinct powers.py
import math
import array
n = 100000
A = array.array('b', [True]*(n+1) )
maxPow = int(math.log(n, 2)) + 1
pows = array.array('i', [0] * maxPow)
for i in range (2, n+1):
    a = i
    for j in range (1,n):
	if(A[a]):
	    A[a] = False
	    pows[j] += 1
	else:
	    if(a>i):
		print(i,j,a)
	    break
	a = a*i
	if (a>n):
	    break

allPows = array.array('b', [True] * (n * (len(pows)-1) + 1) )
total = 0
for i in range (1, len(pows)):
    count=0
    for j in range (2, n+1):
	prod = i*j
	if(allPows[prod]):
	    count += 1
	    allPows[prod] = False
    total += pows[i]*count

print total
    
